homeassistant/components/leneda/coordinator.py

```diff
@@ -0,0 +1,338 @@
+"""The Leneda coordinator."""
+
+from __future__ import annotations
+
+from datetime import datetime, timedelta
+import logging
+import re
+from typing import Any, cast
+
+from leneda import LenedaClient
+from leneda.exceptions import UnauthorizedException
+
+from homeassistant.components.recorder.models import (
+    StatisticData,
+    StatisticMeanType,
+    StatisticMetaData,
+)
+from homeassistant.components.recorder.statistics import (
+    async_add_external_statistics,
+    get_last_statistics,
+    statistics_during_period,
+)
+from homeassistant.components.recorder.util import get_instance
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import UnitOfEnergy
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryAuthFailed
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+from homeassistant.util import dt as dt_util
+
+from .const import (
+    CONF_API_TOKEN,
+    CONF_ENERGY_ID,
+    CONF_METERING_POINTS,
+    DOMAIN,
+    SCAN_INTERVAL,
+    SENSOR_TYPES,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def _create_statistic_id(metering_point: str, obis: str) -> str:
+    """Create a valid statistic ID from metering point and OBIS code."""
+    # Convert to lowercase and replace any non-alphanumeric characters with underscore
+    clean_mp = re.sub(r"[^a-z0-9]", "_", metering_point.lower())
+    clean_obis = re.sub(r"[^a-z0-9]", "_", obis.lower())
+    statistic_id = f"{DOMAIN}:{clean_mp}_{clean_obis}"
+    _LOGGER.debug(
+        "Created statistic ID: %s from metering_point: %s, obis: %s",
+        statistic_id,
+        metering_point,
+        obis,
+    )
+    return statistic_id
+
+
+class LenedaCoordinator(DataUpdateCoordinator[dict[str, dict[str, Any]]]):
+    """Handle fetching Leneda data, updating sensors and inserting statistics."""
+
+    config_entry: ConfigEntry
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        config_entry: ConfigEntry,
+    ) -> None:
+        """Initialize the data handler."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            config_entry=config_entry,
+            name=DOMAIN,
+            update_interval=SCAN_INTERVAL,
+        )
+        self.client = LenedaClient(
+            api_key=config_entry.data[CONF_API_TOKEN],
+            energy_id=config_entry.data[CONF_ENERGY_ID],
+        )
+        self.metering_points = config_entry.data[CONF_METERING_POINTS]
+        self.selected_sensors = config_entry.options.get("selected_sensors", {})
+        _LOGGER.debug(
+            "Initialized coordinator with %s metering points and selected sensors: %s",
+            len(self.metering_points),
+            self.selected_sensors,
+        )
+
+    async def _async_update_data(self) -> dict[str, dict[str, Any]]:
```

Comment:
I'm wondering if updating the statistics is worth seperating out into a seperate PR, there is quite a lot of code in this PR to review.

Also, on the library font, perhaps it could be changed to use aiohttp rather than requests? That way you can call the functions to request data directly rather than needing `async_add_executor_job` jobs.

As a side note - the library doesn't implement a timeout when making a request so there is a potential for this to hang if the server is not responding, or the user doesn't have a stable internet connection. https://github.com/fedus/leneda-client/blob/ddc02c7a4855b28366567e5765c8ab51b66aa6c2/src/leneda/client.py#L92

Reply:
For the library: very good points.

I have updated the library to use async functions and also provided a default 30s connection timeout setting. The integration code has been amended accordingly.